### PR TITLE
Support llama3 autoparallel + pipelining

### PR DESCRIPTION
**Pull request by wconstab**
so far just tested locally
`LOG_RANK=4 CONFIG_FILE=././torchtitan/models/deepseek_v3/train_configs/debug_model.toml ./run_train.sh --model.name llama3_auto_parallel --parallelism.pipeline_parallel_degree 2 --training.steps 100`

Runs and loss converges.

Left one TODO about global-batch-size and gradient accumulation

### CODE REVIEW

File: torchtitan/experiments/auto_parallel/parallelize_llama.py [hunk 1]
```diff
@@ -33,28 +33,65 @@ def parallelize_llama(
...
-                device=torch.device("cuda"),
-            ),
-        )
 
     # TODO make autop work correctly with different combinations of DP, DP+TP, TP, and support DDP / HSDP
     assert parallel_dims.dp_replicate_enabled is False, "DDP not supported yet"
     assert parallel_dims.cp_enabled is False, "CP not supported yet"
-    assert parallel_dims.pp_enabled is False, "PP not supported yet"
 
+    pp_degree = job_config.parallelism.pipeline_parallel_degree
```
Comment (xmfan) [inline torchtitan/experiments/auto_parallel/parallelize_llama.py]: unused pp degree config, should probably raise error when its not local world size

File: torchtitan/experiments/auto_parallel/parallelize_llama.py [hunk 2]
```diff
@@ -33,28 +33,65 @@ def parallelize_llama(
...
+        spmd_dims = []
+        if parallel_dims.dp_replicate_enabled:
+            spmd_dims.append("dp_replicate")
+        if parallel_dims.dp_shard_enabled:
+            spmd_dims.append("dp_shard")
+        if parallel_dims.tp_enabled:
+            spmd_dims.append("tp")
+        spmd_mesh = world_mesh[spmd_dims]
+
+        dp_degree = 1
```
Comment (xmfan) [inline torchtitan/experiments/auto_parallel/parallelize_llama.py]: same, config could specify dp_degree

**Review by xmfan — APPROVED** [review]

File: torchtitan/train.py [hunk 3]
```diff
@@ -492,11 +492,13 @@ def forward_backward_step(
                 )
                 if self.pp_has_first_stage:
                     self.pp_schedule.step(
-                        inputs, target=targets, losses=losses, input_batch=inputs
+                        # TODO: input_batch kwarg only needed for CP, but
+                        # autoparallel doesn't accept kwargs in its forward
+                        inputs, target=targets, losses=losses #, input_batch=inputs
```
Comment (fegin) [inline torchtitan/train.py:497]: Curious, why does CP need `input_batch`?

Comment (wconstab) [reply torchtitan/train.py:497]: I assumed you would know. Am I wrong?

File: torchtitan/experiments/auto_parallel/parallelize_llama.py [hunk 4]
```diff
@@ -33,28 +33,65 @@ def parallelize_llama(
...
-        )
 
     # TODO make autop work correctly with different combinations of DP, DP+TP, TP, and support DDP / HSDP
     assert parallel_dims.dp_replicate_enabled is False, "DDP not supported yet"
     assert parallel_dims.cp_enabled is False, "CP not supported yet"
-    assert parallel_dims.pp_enabled is False, "PP not supported yet"
 
+    pp_degree = job_config.parallelism.pipeline_parallel_degree
+    local_batch_size = job_config.training.local_batch_size
+    spmd_batch_size = local_batch_size
```
Comment (wconstab) [inline torchtitan/experiments/auto_parallel/parallelize_llama.py]: oops this is a bug for the non-pp case.  should be  `local *dp degree` and put in an 'else' branch

File: torchtitan/train.py [hunk 5]
```diff
@@ -492,11 +492,13 @@ def forward_backward_step(
                 )
                 if self.pp_has_first_stage:
                     self.pp_schedule.step(
-                        inputs, target=targets, losses=losses, input_batch=inputs
+                        # TODO: input_batch kwarg only needed for CP, but
+                        # autoparallel doesn't accept kwargs in its forward
```
Comment (ezyang) [inline torchtitan/train.py:496]: Can we just fix this LOL

File: torchtitan/experiments/auto_parallel/parallelize_llama.py [hunk 6]
```diff
@@ -33,28 +33,65 @@ def parallelize_llama(
...
+    logger.info(f"Using {spmd_batch_size=}")
+
+    def input_fn():
+        # TODO(whc) - i am not clear what we put this code here for in the first place
+        # if global_batch_size < 0:
+        #     # This global batch size results in 1 gradient accumulation
+        #     # step.
+        #     dp_degree = parallel_dims.dp_replicate * parallel_dims.dp_shard
+        #     global_batch_size = job_config.training.local_batch_size * dp_degree
+        if parallel_dims.pp_enabled and pp_rank > 0:
```
Comment (ezyang) [inline torchtitan/experiments/auto_parallel/parallelize_llama.py:71]: What a mess. No action here needed, but it's definitely worth thinking about what the terminal UX state here should be.

File: torchtitan/experiments/auto_parallel/parallelize_llama.py (hunk 4 above)
Comment (wconstab) [reply torchtitan/experiments/auto_parallel/parallelize_llama.py]: fixed

File: torchtitan/experiments/auto_parallel/parallelize_llama.py (hunk 1 above)
Comment (wconstab) [reply torchtitan/experiments/auto_parallel/parallelize_llama.py]: i deleted it (it was unused/unneeded). I don't think we need to raise any error.  pp_degree does not need to equal any particular size, and pp can even be disabled.

File: torchtitan/experiments/auto_parallel/parallelize_llama.py (hunk 2 above)
Comment (wconstab) [reply torchtitan/experiments/auto_parallel/parallelize_llama.py]: this is something we could potentially upstream to parallel_dims helper.  cc @tianyu-l any reason not to offer the convenience @property of 'dp_degree' in parallel_dims so you don't have to manually figure out if dp_replicate and/or dp_shard are specified and multiply them?

File: torchtitan/train.py (hunk 3 above)
Comment (fegin) [reply torchtitan/train.py:497]: oh, there was a change to remove the need of `input_batch`. We may want to do the same change to autoparalle.
